html/beps/bep_0046.rst

```diff
@@ -0,0 +1,238 @@
+3. A client may want to be able to make rough download rate estimates of its
+   peers by measuring the timing of HAVE messages
+
+.. _BEP16: http://bittorrent.org/beps/bep_0016.html
+.. _`libtorrent's`: https://github.com/arvidn/libtorrent/commit/66ed31dd4b0c0b6d42c3eeb706477b1b19c8f1ea
+
+In order to not break existing clients, this BEP proposes a flag in the extended
+handshake (BEP10_) to let a peer indicate whether or not it need to receive
+redundant HAVEs.
+
+.. _BEP10: http://bittorrent.org/beps/bep_0010.html
+
+If the key ``rh`` exists and has a value of 0 in the extension handshake, the
+peer indicates that it does *not* need to receive redundant HAVE messages.
+
+If the ``rh`` key exists and is set to 1, the peer explicitly indicates that it
+is interested in redundant HAVE messages.
+
+.. parsed-literal::
+
+	{
+		"m": *extension messages*,
+		**"rh": 0**
+	}
+
+Clients implementing this extension are encouraged to always include this flag
+in their extension handshake, whether it's 0 or 1. This will enable a future
+transition of the default assumption, to be that peers do *not* need redundant
+HAVEs (unless explicitly stated).
+
+All have messages can be considered redundant to a peer that has declared
+itself to be *upload-only* (BEP21_).
+
+A high quality implementation may want to remember HAVE messages that would not
+have been redundant had the peer not been upload-only, and send them if the peer
+changes state to no longer be upload-only. Such bulk update could be compact
+with the extension message proposed in the following section.
+
+.. _BEP21: http://bittorrent.org/beps/bep_0021.html
+
+Compressed bitfields
+====================
+
+This BEP proposes a new extension message which announces multiple pieces in a
+single message in a space-efficient manner. The content of the message is
+conceptually an entire bitfield with bits set for the new pieces the peer has
+received.
+
+Extension message name is "lt_have", as specified in the extension handshake.
+
+This message may substitute the BITFIELD message.
+
+This bitfield is not transferred verbatim, it is *run length encoded*, as a
+sequence of commands. Each command is encoded as a *byte aligned* block. The
+first two bits of each block indicates which command it encodes. That is, the
+most significant bits of the first byte.
+
+These are the commands:
+
+0. fill forward *n* bytes of zeroes (0x00)
+1. fill forward *n* bytes of ones (0xff)
+2. consume the next *n* bytes as verbatim bits of the bitfield
+3. fill forward *n* bytes of zeroes and consume one byte of verbatim bitfield
+
+The start position is always at the first piece in the torrent, position 0. All
+commands progress the cursor forward, towards higher piece indices.
+
+The ``num_bytes`` field in all command blocks are *unsigned* and *big-endian*.
+
+Since there is no point in ever encoding 0 bytes in the ``num_bytes`` field,
+the number of bytes following is ``num_bytes`` + 1. i.e. if the ``num_bytes``
```

Comment:
It seems to me like command 1 is the weakest of these commands, only being useful over command 3 when the run ends on a byte boundary and is followed by an FF byte. You might be able to squeeze a bit more out of the bytes if you used 0 for something else like a "short fill" where ``num_bytes`` is encoded in the same byte instead of requiring an additional byte.

Examples:

```
command 00 (short fill), content 00 (zeroes), bytes 0xA (10+1)
0A => 00 00 00 00 00 00 00 00 00 00 00

command 00 (short fill), content 01 (zeroes), bytes 0x4 (4+1)
14 => FF FF FF FF FF

command 00 (short fill), content 1X (verbatim), bytes 0x3 (3+1)
23 BA AD F0 0D => BA AD F0 0D
```

Reusing your examples might be a bit unfair though, since this command would be optimized for short sequences like them. You'd probably want to run some simulations or grab some real-world data and see how many bytes this could actually save, if any.